Patch author:  Orson Teodoro <orsonteodoro@hotmail.com>
Patch status:  In Development / Testing
Date:  Mon Sep 13 04:26:55 PM PDT 2021 (Unix timestamp: 1631575615)

A problem was encountered when using llvm 13.x corresponding to gentoo's sys-devel/llvm-13.0.0.9999 package.

The resulting output in the following when doing the "llvm-profdata merge" step:

warning: /usr/src/linux/vmlinux.profraw: unsupported instrumentation profile format version
error: no profile can be merged

Modded to included LLVM_INSTR_PROF_RAW_VERSION 6 and 7 support.

The patch status will change once it is confirmed working.

diff -x '*.orig' -x '*.rej' -urpN linux-5.14.3-ot.orig/kernel/pgo/fs.c linux-5.14.3-ot/kernel/pgo/fs.c
--- linux-5.14.3-ot.orig/kernel/pgo/fs.c	2021-09-14 04:25:57.020147052 -0700
+++ linux-5.14.3-ot/kernel/pgo/fs.c	2021-09-14 04:50:28.209111683 -0700
@@ -66,9 +66,16 @@ static void prf_fill_header(void **buffe
 	header->counters_size = prf_cnts_count();
 	header->padding_bytes_after_counters = 0;
 	header->names_size = prf_names_count();
+#if CONFIG_CLANG_VERSION >= 140000 && LLVM_INSTR_PROF_RAW_VERSION >= 7
+	header->counters_delta = (u64)__llvm_prf_cnts_start - (u64)__llvm_prf_data_start;
+#else
 	header->counters_delta = (u64)__llvm_prf_cnts_start;
+#endif
 	header->names_delta = (u64)__llvm_prf_names_start;
 	header->value_kind_last = LLVM_INSTR_PROF_IPVK_LAST;
+#if LLVM_INSTR_PROF_RAW_VERSION >= 6
+	header->binary_id_size = 0; /* optional */
+#endif
 
 	*buffer += sizeof(*header);
 }
diff -x '*.orig' -x '*.rej' -urpN linux-5.14.3-ot.orig/kernel/pgo/Kconfig linux-5.14.3-ot/kernel/pgo/Kconfig
--- linux-5.14.3-ot.orig/kernel/pgo/Kconfig	2021-09-14 04:25:57.016146891 -0700
+++ linux-5.14.3-ot/kernel/pgo/Kconfig	2021-09-14 04:33:09.620506255 -0700
@@ -4,12 +4,18 @@ menu "Profile Guided Optimization (PGO)
 config ARCH_SUPPORTS_PGO_CLANG
 	bool
 
+config PROFRAW_VERSION_OVERRIDE
+	int
+	depends on PROFRAW_VERSION_OVERRIDE_TOGGLE
+	default $(profraw-version)
+
 config PGO_CLANG
 	bool "Enable clang's PGO-based kernel profiling"
 	depends on DEBUG_FS
 	depends on ARCH_SUPPORTS_PGO_CLANG
 	depends on CC_IS_CLANG
 	depends on !ARCH_WANTS_NO_INSTR || CC_HAS_NO_PROFILE_FN_ATTR
+	depends on PROFRAW_VERSION_OVERRIDE_TOGGLE || LLVM_MAIN || LLVM14_INIT || LLVM13_LIVE || LLVM13_RC2 || LLVM13_RC1 || LLVM13_IPR6 || LLVMX_IPR5
 	help
 	  This option enables clang's PGO (Profile Guided Optimization) based
 	  code profiling to better optimize the kernel.
@@ -34,4 +40,41 @@ config PGO_CLANG
 	  Note that the debugfs filesystem has to be mounted to access
 	  profiling data.
 
+choice PGO_CLANG_LLVM_SELECT
+	prompt "LLVM version for profdata compatibility"
+	help
+	  This matches the profdata data structure to a specific version
+	  of LLVM since not enough metadata is provided.
+
+config PROFRAW_VERSION_OVERRIDE_TOGGLE
+	bool "Autodetect profraw version"
+
+config LLVM_MAIN
+	bool "LLVM main branch with the latest commit"
+
+config LLVM14_INIT
+	bool "LLVM 14 init"
+
+config LLVM13_LIVE
+	bool "LLVM 13.x branch with the latest commit"
+
+config LLVM13_RC2
+	bool "LLVM 13 rc2"
+
+config LLVM13_RC1
+	bool "LLVM 13 rc1"
+
+config LLVM13_IPR6
+	bool "LLVM 13 with profraw v6 header from git"
+	help
+	  The build must have both e50a388 and f984ac2 commits applied.
+
+config LLVMX_IPR5
+	bool "LLVM 10 to 13 with profraw v5 and backported commits"
+	help
+	  The required backports of commits 193e41c a63d4f6 must be applied, you
+	  still can do PGO optimization with those versions of LLVM.
+
+endchoice
+
 endmenu
diff -x '*.orig' -x '*.rej' -urpN linux-5.14.3-ot.orig/kernel/pgo/pgo.h linux-5.14.3-ot/kernel/pgo/pgo.h
--- linux-5.14.3-ot.orig/kernel/pgo/pgo.h	2021-09-14 04:25:57.021147092 -0700
+++ linux-5.14.3-ot/kernel/pgo/pgo.h	2021-09-14 05:14:58.064621568 -0700
@@ -43,7 +43,33 @@
 		 (u64)'R' << 8  |	\
 		 (u64)129)
 
-#define LLVM_INSTR_PROF_RAW_VERSION		5
+#if defined(CONFIG_PROFRAW_VERSION_OVERRIDE_TOGGLE)
+#  define LLVM_INSTR_PROF_RAW_VERSION	CONFIG_PROFRAW_VERSION_OVERRIDE
+#  if CONFIG_CLANG_VERSION <= 130000
+#    warning "LLVM needs backports for both 193e41c a63d4f6 commits or update to"
+#    warning "LLVM 13 rc1 or newer"
+#  endif
+#elif defined(CONFIG_LLVM_MAIN) /* corresponds to the main branch */
+#  define LLVM_INSTR_PROF_RAW_VERSION	7
+#elif defined(CONFIG_LLVM14_INIT)
+#  define LLVM_INSTR_PROF_RAW_VERSION	6
+#elif defined(CONFIG_LLVM13_LIVE) /* corresponds to the 13.x branch */
+#  define LLVM_INSTR_PROF_RAW_VERSION	7
+#elif defined(CONFIG_LLVM13_RC2)
+#  define LLVM_INSTR_PROF_RAW_VERSION	7
+#elif defined(CONFIG_LLVM13_RC1)
+#  define LLVM_INSTR_PROF_RAW_VERSION	6
+#elif defined(CONFIG_LLVM13_IPR6)
+#  define LLVM_INSTR_PROF_RAW_VERSION	6
+#elif defined(CONFIG_LLVMX_IPR5)
+#  warning "LLVM needs backports for both 193e41c a63d4f6 commits or update to"
+#  warning "LLVM 13 rc1 or newer"
+#  define LLVM_INSTR_PROF_RAW_VERSION	5
+#else
+#  error "PGO requires >= LLVM 13 rc1 or backports for both 193e41c a63d4f6"
+#  error "commits"
+#endif
+
 #define LLVM_INSTR_PROF_DATA_ALIGNMENT		8
 #define LLVM_INSTR_PROF_IPVK_FIRST		0
 #define LLVM_INSTR_PROF_IPVK_LAST		1
@@ -72,6 +100,11 @@
 struct llvm_prf_header {
 	u64 magic;
 	u64 version;
+#if defined(LLVM_MAIN) \
+	|| ( LLVM_INSTR_PROF_RAW_VERSION >= 7 \
+		&& CONFIG_CLANG_VERSION == 140000 )
+	u64 binary_id_size;
+#endif
 	u64 data_size;
 	u64 padding_bytes_before_counters;
 	u64 counters_size;
@@ -80,6 +113,15 @@ struct llvm_prf_header {
 	u64 counters_delta;
 	u64 names_delta;
 	u64 value_kind_last;
+#if defined(CONFIG_LLVM14_INIT) \
+	|| defined(CONFIG_LLVM13_LIVE) \
+	|| defined(CONFIG_LLVM13_RC2) \
+	|| defined(CONFIG_LLVM13_RC1) \
+	|| defined(CONFIG_LLVM13_IPR6) \
+	|| ( LLVM_INSTR_PROF_RAW_VERSION >= 6 \
+		&& CONFIG_CLANG_VERSION < 140000 )
+	u64 binary_id_size;
+#endif
 };
 
 /**
diff -x '*.orig' -x '*.rej' -urpN linux-5.14.3-ot.orig/scripts/Kconfig.include linux-5.14.3-ot/scripts/Kconfig.include
--- linux-5.14.3-ot.orig/scripts/Kconfig.include	2021-08-29 15:04:50.000000000 -0700
+++ linux-5.14.3-ot/scripts/Kconfig.include	2021-09-14 04:35:10.483354430 -0700
@@ -63,3 +63,7 @@ ld-version := $(shell,set -- $(ld-info)
 cc-option-bit = $(if-success,$(CC) -Werror $(1) -E -x c /dev/null -o /dev/null,$(1))
 m32-flag := $(cc-option-bit,-m32)
 m64-flag := $(cc-option-bit,-m64)
+
+# Get the profraw version
+$(warning-if,$(failure,which clang++),Did you add the llvm bin path to the PATH variable?)
+profraw-version := $(shell,$(srctree)/scripts/profraw-version.sh || echo "5")
diff -x '*.orig' -x '*.rej' -urpN linux-5.14.3-ot.orig/scripts/profraw-check.cc linux-5.14.3-ot/scripts/profraw-check.cc
--- linux-5.14.3-ot.orig/scripts/profraw-check.cc	1969-12-31 16:00:00.000000000 -0800
+++ linux-5.14.3-ot/scripts/profraw-check.cc	2021-09-14 04:30:03.551038772 -0700
@@ -0,0 +1,7 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT OR Apache-2.0
+#include <llvm/ProfileData/InstrProf.h>
+#include <iostream>
+
+int main(int argc, char *argv[]) {
+	std::cout << llvm::RawInstrProf::Version;
+}
diff -x '*.orig' -x '*.rej' -urpN linux-5.14.3-ot.orig/scripts/profraw-version.sh linux-5.14.3-ot/scripts/profraw-version.sh
--- linux-5.14.3-ot.orig/scripts/profraw-version.sh	1969-12-31 16:00:00.000000000 -0800
+++ linux-5.14.3-ot/scripts/profraw-version.sh	2021-09-14 05:08:44.118796134 -0700
@@ -0,0 +1,23 @@
+#!/bin/bash
+# SPDX-License-Identifier: GPL-2.0-or-later OR MIT OR Apache-2.0
+
+PROFRAW_VERSION_MIN=5
+
+error() {
+	exit 1
+}
+
+get_profraw_version() {
+	which clang++ 2>/dev/null 1>/dev/null || error
+	which llvm-config 2>/dev/null 1>/dev/null || error
+	S=$(dirname "$BASH_SOURCE")
+	cd "${S}" 2>/dev/null 1>/dev/null || error
+	rm profraw-check.o profraw-check 2>/dev/null
+	clang++ -c $(llvm-config --cxxflags) -o profraw-check.o profraw-check.cc \
+		2>/dev/null 1>/dev/null || error
+	clang++ $(llvm-config --ldflags --libs) -o profraw-check profraw-check.o \
+		2>/dev/null 1>/dev/null || error
+	./profraw-check || error
+}
+
+get_profraw_version
